File: Tutorials/binomial_tree.py
import numpy as np

def binomial_tree(S0, K, T, r, sigma, num_steps, is_call=True):
    """
    Price an option using the binomial tree method.

    Parameters:
    - S0: Initial stock price (float)
    - K: Strike price (float)
    - T: Time to maturity (float, in years)
    - r: Risk-free rate (float)
    - sigma: Volatility (float)
    - num_steps: Number of steps in the binomial tree (int)
    - is_call: Whether the option is a call (default True). False for a put option.

    Returns:
    - option_price: float
    """
    delta_t = T / num_steps
    u = np.exp(sigma * np.sqrt(delta_t))  # Up factor
    d = 1 / u  # Down factor
    p = (np.exp(r * delta_t) - d) / (u - d)  # Risk-neutral probability
    q = 1 - p  # Complement probability

    # Initialize asset price tree
    S = np.zeros((num_steps + 1, num_steps + 1))
    S[0, 0] = S0
    for i in range(1, num_steps + 1):
        S[i, 0] = S[i - 1, 0] * u
        for j in range(1, i + 1):
            S[i, j] = S[i - 1, j - 1] * d

    # Initialize option value tree
    V = np.zeros((num_steps + 1, num_steps + 1))
    for j in range(num_steps + 1):
        if is_call:
            V[num_steps, j] = max(S[num_steps, j] - K, 0)  # Call option payoff
        else:
            V[num_steps, j] = max(K - S[num_steps, j], 0)  # Put option payoff

    # Backward induction
    for i in range(num_steps - 1, -1, -1):
        for j in range(i + 1):
            V[i, j] = np.exp(-r * delta_t) * (p * V[i + 1, j] + q * V[i + 1, j + 1])

    return V[0, 0]

# Example usage
if __name__ == "__main__":
    num_steps = 200
    S0 = 10
    K = 10
    sigma = 0.4
    r = 0.05
    T = 1

    call_price = binomial_tree(S0, K, T, r, sigma, num_steps, is_call=True)
    put_price = binomial_tree(S0, K, T, r, sigma, num_steps, is_call=False)
    print(f"Call Option Price: {call_price:.6f}")
    print(f"Put Option Price: {put_price:.6f}")

File: OptionPricingFramework/src/methods/BinomialTree.cpp
#include "../../include/Methods/BinomialTree.hpp"
#include <torch/torch.h>
#include <cmath>

BinomialTree::BinomialTree(torch::Tensor S0, torch::Tensor K, torch::Tensor r, torch::Tensor sigma, torch::Tensor T, int N)
    : S0(S0), K(K), r(r), sigma(sigma), T(T), N(N) {}

torch::Tensor BinomialTree::price()
{
    return priceOption();
}

torch::Tensor BinomialTree::priceOption()
{
    torch::Tensor dt = T / N;
    torch::Tensor u = torch::exp(sigma * torch::sqrt(dt));
    torch::Tensor d = torch::exp(-sigma * torch::sqrt(dt));
    torch::Tensor p = (torch::exp(r * dt) - d) / (u - d);

    // Ensure dtype and device consistency
    torch::Tensor optionPrice = torch::zeros({N + 1}, S0.options());

    // Terminal payoff at maturity
    for (int i = 0; i <= N; ++i)
    {
        torch::Tensor stockPrice = S0 * torch::pow(u, N - i) * torch::pow(d, i);
        optionPrice[i] = torch::max(stockPrice - K, torch::tensor(0.0, S0.options()));
    }

    // Backward induction
    for (int j = N - 1; j >= 0; --j)
    {
        torch::Tensor newOptionPrice = torch::zeros({j + 1}, S0.options());
        for (int i = 0; i <= j; ++i)
        {
            newOptionPrice[i] = (p * optionPrice[i] + (1 - p) * optionPrice[i + 1]) * torch::exp(-r * dt);
        }
        optionPrice = newOptionPrice;
    }

    return optionPrice[0];
}
